nncf/common/pruning/schedulers.py

```diff
@@ -10,112 +10,146 @@
 
-        self.algo = pruning_algo
+    A pruning scheduler is an object which specifies the pruning
+    level at each training epoch. It involves a scheduling algorithm,
+    defined in the `_calculate_pruning_level()` method and a state
+    (some parameters required for current pruning level calculation)
+    defined in the `__init__()` method.
+    """
 
-        # Number of initial steps of training before pruning
-        self.num_init_steps = self._params.get('num_init_steps', 0)
-        self.pruning_steps = self._params.get('pruning_steps', 100)
+    def __init__(self, controller, params: dict):
+        """
+        Initializes the internal state of the pruning scheduler specified by:
+            - controller: Pruning algorithm controller.
+            - initial_pruning: Pruning level which already has been
+                applied to the model. It is the level at which the schedule begins.
+            - target_pruning: Pruning level at which the schedule ends.
+            - num_warmup_epochs: Number of epochs for model pre-training before pruning.
+            - num_pruning_epochs: Number of epochs during which the pruning level
+                is increased from `initial_pruning` to `target_pruning`.
+            - freeze_epoch: Zero-based index of the epoch from which the pruning
+                mask will be frozen and will not be trained.
+            - current_pruning: Pruning level for the `current_epoch`.
+
+        :param controller: Pruning algorithm controller.
+        :param params: Parameters of the scheduler.
+        """
+        super().__init__()
+        self.controller = controller
+        self.initial_pruning = self.controller.pruning_init
 
-        # Pruning rates
-        self.initial_pruning = self.algo.pruning_init
-        if self.algo.prune_flops:
-            self.pruning_target = self._params.get('pruning_flops_target')
+        if self.controller.prune_flops:
+            self.target_pruning = params.get('pruning_flops_target')
         else:
-            self.pruning_target = self._params.get('pruning_target', 0.5)
+            self.target_pruning = params.get('pruning_target', 0.5)
 
-    def load_state(self, state):
-        super().load_state(state)
-        self._set_pruning_level()
+        self.num_warmup_epochs = params.get('num_init_steps', 0)
+        self.num_pruning_epochs = params.get('pruning_steps', 100)
+        self.freeze_epoch = self.num_warmup_epochs + self.num_pruning_epochs
+        self.current_pruning = self.initial_pruning
 
-    def epoch_step(self, next_epoch=None):
-        super().epoch_step(next_epoch)
-        self._set_pruning_level()
+    def _maybe_freeze(self) -> None:
+        """
+        Checks if pruning mask needs to be frozen and not to be trained.
+        If freezing is needed, then the internal state of the `controller`
+        object will be changed.
+        """
+        if self.current_epoch >= self.freeze_epoch:
+            self.controller.freeze()
```

Comment:
You saved a single line of code at a high cost of code readability. Suggest explicitly inlining these function calls.

Reply:
I agree with you. I fixed it.